hoon/lib/jock.hoon

```diff
@@ -553,33 +609,59 @@
     =^  inner-two  tokens
       (match-inner-jock tokens)
     [[%compare [%limb limbs] comparator inner-two] tokens]
-  ?:  (has-punctuator -.tokens %'((')
-    |-
-    =.  tokens  +.tokens
+  ::  - %call ('((' is the next token)
+  ?:  |((has-punctuator -.tokens %'((') (has-punctuator -.tokens %'('))
+    =?  tokens  ?=(%'((' ->.tokens)  [[%punctuator %'('] +.tokens]
     =^  arg  tokens
-      (match-inner-jock tokens)
-    ?>  (got-punctuator -.tokens %')')
+      (match-pair-inner-jock tokens)
     ::  TODO: check if we're in a compare
-    [[%call [%limb limbs] `arg] +.tokens]
+    ?:  (is-type name)
+      :: inject constructor call
+      [[%call [%limb [[%name p=%new] limbs]] `arg] tokens]
+    [[%call [%limb limbs] `arg] tokens]
   [[%limb limbs] tokens]
 ::
+++  make-jlimb
+  |=  name=cord
+  ^-  jlimb
+  ?:  ((sane %tas) name)
+    [%name name]
+  [%type name]
+::
 ::  Metatype is a container type like List or Set
 ++  match-metatype
   |=  =tokens
   ^-  [jype (list token)]
   ?:  =(~ tokens)  ~|("expect expression starting with type. token: ~" !!)
   ?:  !=(%type -<.tokens)
     ~|("expect type. token: {<-.tokens>}" !!)
-  =/  type
-    ?:  =([%type 'List'] -.tokens)
-      %list
-    ?:  =([%type 'Set'] -.tokens)
-      %set
-    !!  ::  TODO generalize
+  =/  type=cord
+    ?:  =([%type 'List'] -.tokens)  %list
```

Comment:
Could these special cases be removed

Reply:
Not in this branch yet. There will probably always be special-casing in some form because we have native syntax.